src/algo/depth-to-rgb-calibration/optimizer.cpp

```diff
@@ -0,0 +1,1720 @@
+    
+    calib old_calib = decompose( curr_params.curr_p_mat, _original_calibration );
+    auto uvmap_old = get_texture_map(new_vertices, old_calib, curr_params.curr_p_mat );
+    //curr_params.cost = calc_cost( z_data, yuy_data, uvmap_old );
+
+    calib new_calib = decompose( new_params.curr_p_mat, _original_calibration );
+    auto uvmap_new = get_texture_map(new_vertices, new_calib, new_params.curr_p_mat );
+    new_params.cost = calc_cost( _z, _yuy, uvmap_new );
+
+    auto diff = calc_cost_per_vertex_diff( _z, _yuy, uvmap_old, uvmap_new );
+
+    auto iter_count = 0;
+    while( diff >= step_size * t
+           && abs( step_size ) > _params.min_step_size
+           && iter_count++ < _params.max_back_track_iters )
+    {
+        AC_LOG( DEBUG, "    back tracking line search cost= " << AC_D_PREC << new_params.cost );
+        step_size = _params.tau * step_size;
+
+        new_params.curr_p_mat = curr_params.curr_p_mat + unit_grad * step_size;
+        
+        new_calib = decompose( new_params.curr_p_mat, _original_calibration );
+        uvmap_new = get_texture_map(new_vertices, new_calib, new_params.curr_p_mat);
+        new_params.cost = calc_cost( _z, _yuy, uvmap_new);
+        diff = calc_cost_per_vertex_diff( _z, _yuy, uvmap_old, uvmap_new );
+    }
+
+    if(diff >= step_size * t )
+    {
+        new_params = curr_params;
+    }
+
+    if (data)
+    {
+        data->iteration_data_p.grads_norma = curr_params.calib_gradients.get_norma();
+        data->iteration_data_p.grads_norm = grads_over_norm;
+        data->iteration_data_p.normalized_grads = grad;
+        data->iteration_data_p.unit_grad = unit_grad;
+        data->iteration_data_p.back_tracking_line_search_iters = iter_count;
+        data->iteration_data_p.t = t;
+    }
+    return new_params;
+}
+
+void optimizer::set_final_data(const std::vector<double3>& vertices,
+    const p_matrix& p_mat,
+    const p_matrix& p_mat_opt)
+{
+    _vertices_from_bin = vertices;
+    _p_mat_from_bin = p_mat;
+    _p_mat_from_bin_opt = p_mat_opt;
+}
+
+void optimizer::set_cycle_data(const std::vector<double3>& vertices, 
+    const rs2_intrinsics_double& k_depth,
+    const p_matrix& p_mat, 
+    const algo_calibration_registers& dsm_regs_cand,
+    const rs2_dsm_params_double& dsm_params_cand)
+{
+    _vertices_from_bin = vertices;
+    _k_dapth_from_bin = k_depth;
+    _p_mat_from_bin = p_mat;
+    _dsm_regs_cand_from_bin = dsm_regs_cand;
+    _dsm_params_cand_from_bin = dsm_params_cand;
+}
+
+size_t optimizer::optimize_p
+(
+    const optimization_params& params_curr,
+    const std::vector<double3>& new_vertices,
+    optimization_params& params_new,
+    calib& optimaized_calibration,
+    calib& new_rgb_calib_for_k_to_dsm,
+    rs2_intrinsics_double& new_z_k,
+    std::function<void(data_collect const&data)> cb,
+    data_collect* data 
+)
+{
+
+    size_t n_iterations = 0;
+    auto curr = params_curr;
+    while (1)
```

Comment:
What if optimizer doesn't converge - is there a timeout/max iterations threshold to handle this ?

Reply:
No timeout, but iterations: yes. See line 1585.